### PR TITLE
Run synchronize in a single transaction per connection

### PR DESCRIPTION
**Pull request by kabirbaidhya**
### Changes
 - Run the entire process in a single transaction per database connection.
 - Handle errors gracefully for each transaction - even while error occurred in one connection, the process for other connections can still continue.
 - Errors are displayed at the end after once all completed.
 - Better visualization of final result. It displays how many were successful and how many failed. 
 - **Tested** 
    - with single connection
    - with multiple connections both successful and failing. 
    - Experimental versions used for testing v1.0.0-experimental.9-2
 - **Bonus**
   * Display elapsed time for each connection. 

### Screenshots

**1. Successful Output (2 connections, mssql)**
Selection_109

**2. Failing Output (2 connections, mssql)**
Selection_110

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -47,14 +47,27 @@ export async function resolveConnections(): Promise<ConnectionConfig[]> {
...
   return result;
 }
 
+/**
+ * Get the connection id from the config.
+ *
+ * @param {ConnectionConfig} connectionConfig
+ * @returns {string}
+ */
+export function getConnectionId(connectionConfig: ConnectionConfig): string {
```
Comment (kabirbaidhya) [inline src/config.ts]: Add tests for this.

File: src/api.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ */
+function mapToConnectionReferences(connectionList: (ConnectionConfig | Knex)[]): ConnectionReference[] {
+  return connectionList.map(connection => {
+    if (isKnexInstance(connection)) {
+      log(`Received connection instance to database: ${connection.client.config.connection.database}`);
+
+      // TODO: Ask for `id` explicitly in for programmatic API,
+      // when Knex instance is passed directly.
+      // This implies a breaking change with the programmatic API.
+      return { connection, id: undefined };
```
Comment (SafalPandey) [inline src/api.ts]:
```suggestion
      return { connection, id: getConnectionId(getConfig(connection)) };
```

**Comment by kabirbaidhya** [comment]
@mesaugat @SafalPandey @cham11ng This is now fully tested and ready for review. :crossed_fingers:

File: src/SyncDb.ts [hunk 3]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
       process.exit(-1);
     }
   }
+
+  /**
+   * Display the result.
+   *
+   * @param {SyncResult[]} result
+   */
+  async displayResult(result: SyncResult[]) {
```
Comment (mesaugat) [inline src/SyncDb.ts]:
```suggestion
  async displayResult(results: SyncResult[]) {
```

File: src/SyncDb.ts [hunk 4]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
+      await printError(e.toString());
+
       process.exit(-1);
     }
   }
+
+  /**
+   * Display the result.
+   *
+   * @param {SyncResult[]} result
```
Comment (mesaugat) [inline src/SyncDb.ts]: `results`

File: src/SyncDb.ts [hunk 5]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
+   */
+  async displayResult(result: SyncResult[]) {
+    const failed = result.filter(attempt => !attempt.success);
```
Comment (mesaugat) [inline src/SyncDb.ts]: Positive cases are easier to read.
 
```suggestion
    const success = results.filter(result => result.success);
```

File: src/SyncDb.ts [hunk 6]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
+      // Display output.
+      await printLine(`Synchronization successful for ${successfulCount} / ${totalCount} connection(s).`);
+    }
+
+    if (failed.length === 0) {
+      return process.exit(0);
+    }
+
+    // Display all errors.
+    await printLine(`Failed for following ${failedCount} connection(s):\n`);
```
Comment (mesaugat) [inline src/SyncDb.ts]:
```suggestion
    await printLine(`Synchronization failed for ${failedCount} connection(s):\n`);
```

File: src/SyncDb.ts [hunk 7]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
+
+    failed.forEach(async (attempt, index) => {
+      await printLine(`${index + 1}) ${attempt.connectionId}`);
+      await printError(attempt.error);
+      await printLine();
+    });
+
+    throw new Error(`Synchronization failed for ${failedCount} / ${totalCount} connections.`);
```
Comment (mesaugat) [inline src/SyncDb.ts]: Why throw at the end? Displaying all the errors is not enough?

Comment (mesaugat) [inline src/SyncDb.ts]: Why throw at the end? Displaying all the errors is not enough?

File: src/SyncDb.ts [hunk 8]
```diff
@@ -20,35 +24,93 @@ class SyncDb extends Command {
...
+    if (failed.length === 0) {
+      return process.exit(0);
+    }
+
+    // Display all errors.
+    await printLine(`Failed for following ${failedCount} connection(s):\n`);
+
+    failed.forEach(async (attempt, index) => {
+      await printLine(`${index + 1}) ${attempt.connectionId}`);
+      await printError(attempt.error);
```
Comment (mesaugat) [inline src/SyncDb.ts]: Ideally the best way is to only show the message for the error and not the whole stack trace. The whole stack trace can be shown in debug mode.

File: src/api.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  config: SyncConfig,
+  conn: ConnectionConfig[] | ConnectionConfig | Knex[] | Knex,
+  options?: SyncParams
+): Promise<SyncResult[]> {
+  log('Starting to synchronize.');
+
+  const connectionList = Array.isArray(conn) ? conn : [conn];
+  const connections = mapToConnectionReferences(connectionList);
+  const params = mergeDeepRight(DEFAULT_SYNC_PARAMS, options);
+  const cliEnvironment = process.env.SYNC_DB_CLI === 'true';
```
Comment (mesaugat) [inline src/api.ts]:
```suggestion
  const isCLI = process.env.SYNC_DB_CLI === 'true';
```

File: src/api.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  const promises = connections.map(({ connection, id: connectionId }) =>
+    synchronizeDatabase(connection, {
+      config,
+      params,
+      connectionId,
+      cliEnvironment
+    })
+  );
+
+  const result = await Promise.all(promises);
```
Comment (mesaugat) [inline src/api.ts]:
```suggestion
  const results = await Promise.all(promises);
```

File: src/api.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+  log('Finished all');
+
+  return result;
```
Comment (mesaugat) [inline src/api.ts]: .

File: src/api.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      config,
+      params,
+      connectionId,
+      cliEnvironment
+    })
+  );
+
+  const result = await Promise.all(promises);
+
+  log('Finished all');
```
Comment (mesaugat) [inline src/api.ts]:
```suggestion
  log('Synchronization completed.');
```

File: src/SyncDb.ts (hunk 5 above)
Comment (kabirbaidhya) [reply src/SyncDb.ts]: Yes, but we need all the failed attempts instead. For successful ones only the counts matter - for error cases we need full error trace. 

```ts
    failed.forEach(async (attempt, index) => {
      await printLine(`${index + 1}) ${attempt.connectionId}`);
      await printError(attempt.error);
      await printLine();
    });
```

File: src/api.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+function mapToConnectionReferences(connectionList: (ConnectionConfig | Knex)[]): ConnectionReference[] {
+  return connectionList
+    .map(connection => {
+      if (isKnexInstance(connection)) {
+        log(`Received connection instance to database: ${connection.client.config.connection.database}`);
+
+        // TODO: Ask for `id` explicitly in for programmatic API,
+        // when Knex instance is passed directly.
+        // This implies a breaking change with the programmatic API.
+        return { connection, id: undefined };
```
Comment (mesaugat) [inline src/api.ts]: Not sure if I understand correctly but do we need the second `map`?

```js
return { connection, id: getConnectionId(getConfig(connection)) }
```

File: src/domain/SyncContext.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import SyncConfig from './SyncConfig';
+import SyncParams from './SyncParams';
+
+/**
+ * Synchronize context parameters for the current database connection.
+ */
+interface SyncContext {
+  config: SyncConfig;
+  connectionId: string;
+  cliEnvironment: boolean;
```
Comment (mesaugat) [inline src/domain/SyncContext.ts]: .

File: src/services/dbSyncer.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
+import * as Knex from 'knex';
```
Comment (mesaugat) [inline src/services/dbSyncer.ts]: Filename could still be `dbSync`. Haha `dbSyncer`?

Comment (kabirbaidhya) [reply src/services/dbSyncer.ts]: Naming is always difficult :slightly_smiling_face:

File: src/SyncDb.ts (hunk 7 above)
Comment (SafalPandey) [reply src/SyncDb.ts]: We should throw the error to make sure the program exits with error code other than 0.

File: src/api.ts (hunk 13 above)
Comment (kabirbaidhya) [reply src/api.ts]: Both are user provided values so we'll never be sure if we get the `id` out in both cases. Here we won't get it anyway so it's `undefined` to explicitly say so. 

If we do it here - then we'll need to do it in two places - that's why I moved it to a second map - this ensures all the connections has id (or has a fallback id if not provided)

File: src/api.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+      }
+
+      log(`Creating a connection to database: ${connection.host}/${connection.database}`);
+
+      return { connection: createInstance(connection), id: connection.id };
```
Comment (kabirbaidhya) [inline src/api.ts]: @mesaugat There's no assurance that `id` has a value. This is just a typecheck not a run-time assurance.

File: src/SyncDb.ts (hunk 7 above)
Comment (kabirbaidhya) [reply src/SyncDb.ts]: Yes, or else we'll need to exit non-zero in multiple places which isn't good either.

File: src/SyncDb.ts (hunk 8 above)
Comment (kabirbaidhya) [reply src/SyncDb.ts]: Make sense. I'll send out the stack trace in `DEBUG` mode.

**Comment by kabirbaidhya** [comment]
@mesaugat This PR is a small one - which could be merged first and I can rebase it here before making changes in that utility.

File: src/SyncDb.ts (hunk 7 above)
Comment (mesaugat) [reply src/SyncDb.ts]: So `process.exit(1)`

File: src/api.ts (hunk 13 above)
Comment (mesaugat) [reply src/api.ts]: I still feel doing the same in two place is better than using a `map`. Don't need to be DRY everywhere and is a lot easier to understand. 🤷‍♂

File: src/SyncDb.ts (hunk 7 above)
Comment (kabirbaidhya) [reply src/SyncDb.ts]: It's not a good idea of exit in multiple places - throwing exceptions is the right way to raise errors. We just need to ensure that there's a root handler which catches all of them.

Comment (mesaugat) [reply src/SyncDb.ts]: Cool. We already have a try catch in place. 

My only issue is why does a function that simply displays a summary or a result needs to throw any kind of error.

Comment (kabirbaidhya) [reply src/SyncDb.ts]: There's a single place where all exceptions are caught and a non zero exit code is thrown in case of errors.

Comment (kabirbaidhya) [reply src/SyncDb.ts]: Good point. This doesn't sound right with the display result function as it says.

File: src/SyncDb.ts [hunk 17]
```diff
@@ -20,35 +24,98 @@ class SyncDb extends Command {
...
   }
+
+  /**
+   * Display the results.
+   *
+   * @param {SyncResult[]} results
+   */
+  async displayResult(results: SyncResult[]) {
+    const totalCount = results.length;
+    const failedAttempts = results.filter(attempt => !attempt.success);
```
Comment (mesaugat) [inline src/SyncDb.ts]:
```suggestion
    const failedAttempts = results.filter(result => !result.success);
```

File: src/SyncDb.ts (hunk 7 above)
Comment (mesaugat) [reply src/SyncDb.ts]: The problem is that `displayResult` did not fail but `synchronize` or something internally did. Right now display result is trying to say that something failed upstream thus I should fail. When everything is well, `displayResult` does what it is supposed to do. It prints the result and that is what it should do.

The throw should happen after `displayResult`.

```js
      const results = await synchronize(config, connections, params);

      await this.displayResult(results);

      const fail = results.find(result => result.success === false)
      
      if (fail) {
         throw new Error('...');
      }
    } catch (e) {
      // Send verbose error with stack trace to debug logs.
      log(e);

      await printError(e.toString());

      process.exit(-1);
    }
```

File: src/SyncDb.ts [hunk 18]
```diff
@@ -20,35 +24,104 @@ class SyncDb extends Command {
...
+      const { synchronize } = await import('./api');
+
+      await printLine('Synchronizing...\n');
 
-      await synchronize(config, connections, params);
-      process.exit(0);
+      const results = await synchronize(config, connections, params);
+      const { totalCount, failedCount } = await this.processResults(results);
+
+      if (failedCount > 0) {
```
Comment (mesaugat) [inline src/SyncDb.ts]: Haha, I made a similar comment and you pushed it as well.

Comment (mesaugat) [reply src/SyncDb.ts]: https://github.com/leapfrogtechnology/sync-db/pull/39#discussion_r384325669

**Pull request merged**